### PR TITLE
Feature/click action

### PR DESCRIPTION
**Pull request by Andries-Smit**
Execute a microflow or open a page when an image is clicked

### CODE REVIEW

File: src/com/mendix/widget/carousel/Carousel.ts [hunk 1]
```diff
@@ -10,8 +10,13 @@ class CarouselDojo extends WidgetBase {
...
 
-    postCreate() {
-        render(createElement(Carousel, { images: this.staticImages }), this.domNode);
+    update(contextObject: mendix.lib.MxObject, callback?: Function) {
+        render(createElement(Carousel, {
+            contextGuid: contextObject ? contextObject.getGuid() : null,
+            images: this.staticImages
+        }), this.domNode);
+
+        if (callback) { callback(); }
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/Carousel.ts]: Our codestyle is either:
`if (callback) callback();`
or
```
if (callback) {
    callback();
}
```

File: src/com/mendix/widget/carousel/Carousel.xml [hunk 2]
```diff
@@ -14,6 +14,17 @@
                     <category>Image</category>
                     <description>Image to add to the carousel</description>
                 </property>
+                <property key="onClickMicroflow" type="microflow" required="false">
+                    <caption>Microflow</caption>
+                    <category>On click</category>
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/Carousel.xml]: On click action?

File: src/com/mendix/widget/carousel/components/Carousel.ts [hunk 3]
```diff
@@ -68,4 +72,21 @@ export class Carousel extends Component<CarouselProps, { activeIndex: number }>
 
         this.setState({ activeIndex: newActiveIndex });
     }
+
+    private executeAction(microflow?: string, form?: string) {
+        if (microflow) {
+            window.mx.ui.action(microflow, {
+                error: (error: Error) =>
+                    window.mx.ui.error(`An error occurred while executing action: ${error.message}`, true),
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/Carousel.ts]: This is pretty user unfriendly. Why not render a message a `<p class="has-error">` or something? On retry you can remove it again.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts [hunk 4]
```diff
@@ -152,4 +161,104 @@ describe("Carousel", () => {
...
+
+    describe("with click action", () => {
+        it("executes the specified microflow when a carousel item is clicked", () => {
+            images = [ { onClickMicroflow: lorem.word(), url: image.imageUrl() } ];
+            spyOn(window.mx.ui, "action").and.callThrough();
+            carousel = shallow(createElement(Carousel, { images, contextGuid }));
+
+            carousel.find(CarouselItem).simulate("click");
+
+            expect(window.mx.ui.action).toHaveBeenCalledWith( images[0].onClickMicroflow, {
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: There's an extra space between `(` and `images[0]`.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts [hunk 5]
```diff
@@ -152,4 +161,104 @@ describe("Carousel", () => {
...
+            });
+        });
+
+        it("shows an error when a carousel item is clicked with an invalid microflow", () => {
+            const invalidAction = "invalid_action";
+            images = [ { onClickMicroflow: invalidAction, url: image.imageUrl() } ];
+            const actionErrorMessage = "An error occurred while executing action: mx.ui.action error mock";
+
+            spyOn(window.mx.ui, "action").and.callFake((actionname: string, action: { error: (e: Error) => void}) => {
+                if (actionname === invalidAction) {
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Why check for an invalid action? Just always return an error in this case, it only applies to this test.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts [hunk 6]
```diff
@@ -152,4 +161,104 @@ describe("Carousel", () => {
             expect(carouselItems.at(1).props().active).toBe(true);
         });
     });
+
+    describe("with click action", () => {
+        it("executes the specified microflow when a carousel item is clicked", () => {
+            images = [ { onClickMicroflow: lorem.word(), url: image.imageUrl() } ];
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: You are doing the same setup over and over again. Why not extract it to a function which returns a `ShallowWrapper`? That would make your tests more readable.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts [hunk 7]
```diff
@@ -152,4 +161,104 @@ describe("Carousel", () => {
...
+            expect(window.mx.ui.openForm).toHaveBeenCalledWith(images[0].onClickForm, { error: jasmine.any(Function) });
+        });
+
+        it("shows an error when a carousel item is clicked with an invalid form", () => {
+            const invalidForm = "invalid_form";
+            images = [ { onClickForm: invalidForm, url: image.imageUrl() } ];
+            const openFormErrorMessage = "An error occurred while opening form: mx.ui.openForm error mock";
+
+            spyOn(window.mx.ui, "openForm").and.callFake((path: string, args: { error: (e: Error) => void}) => {
+                if (path === invalidForm) {
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Same comment as above.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts [hunk 8]
```diff
@@ -152,4 +161,104 @@ describe("Carousel", () => {
...
+            });
+            spyOn(window.mx.ui, "error").and.callThrough();
+            carousel = shallow(createElement(Carousel, { images, contextGuid: random.uuid() }));
+
+            carousel.find(CarouselItem).simulate("click");
+
+            expect(window.mx.ui.error).toHaveBeenCalledWith(openFormErrorMessage, true);
+        });
+
+        it("executes the microflow with both microflow and form are specified, on carousel item is clicked", () => {
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: => "with both microflow and form specified"

File: src/com/mendix/widget/carousel/components/__tests__/CarouselItem.spec.ts [hunk 9]
```diff
@@ -44,4 +44,20 @@ describe("CarouselItem", () => {
         expect(carouselItem.instance().props.active).toBe(false);
         expect(carouselItem.hasClass("active")).toBe(false);
     });
+
+    describe("image", () => {
+        let clickCount = 0;
```
Comment (johanflint) [inline src/com/mendix/widget/carousel/components/__tests__/CarouselItem.spec.ts]: Don't use a counter, use a spy and verify that it has been called. More readable and more descriptive.

File: tests/matchers.ts [hunk 10]
```diff
@@ -14,41 +13,39 @@ function compareTrees(actual: any, expected: ReactChild, strict: boolean) {
 
 function toBeElement() {
     return {
-        compare: function(actual: ReactChild, expected: ReactChild) {
-            return compareTrees(actual, expected, true);
-        }
+        compare: (actual: ReactChild, expected: ReactChild) =>
```
Comment (johanflint) [inline tests/matchers.ts]: Just put it on the same line :)

File: tests/matchers.ts [hunk 11]
```diff
@@ -14,41 +13,39 @@ function compareTrees(actual: any, expected: ReactChild, strict: boolean) {
...
+            compareTrees(actual, expected, true)
     };
 }
 
 function toMatchStructure() {
     return {
-        compare: function(actual: HTMLElement, expected: ReactChild) {
-            return compareTrees(actual, expected, false);
-        }
+        compare: (actual: HTMLElement, expected: ReactChild) =>
```
Comment (johanflint) [inline tests/matchers.ts]: And here.

File: src/com/mendix/widget/carousel/Carousel.ts (hunk 1 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/Carousel.ts]: Gone with the former.

File: src/com/mendix/widget/carousel/Carousel.xml (hunk 2 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/Carousel.xml]: Done.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts (hunk 4 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Removed.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts (hunk 8 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Done.

File: tests/matchers.ts (hunk 10 above)
Comment (edwinmp) [reply tests/matchers.ts]: Done.

File: tests/matchers.ts (hunk 11 above)
Comment (edwinmp) [reply tests/matchers.ts]: Done.

File: src/com/mendix/widget/carousel/components/__tests__/CarouselItem.spec.ts (hunk 9 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/CarouselItem.spec.ts]: Done.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts (hunk 6 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Done.

File: src/com/mendix/widget/carousel/components/Carousel.ts (hunk 3 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/Carousel.ts]: Done.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts (hunk 5 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Done.

File: src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts (hunk 7 above)
Comment (edwinmp) [reply src/com/mendix/widget/carousel/components/__tests__/Carousel.spec.ts]: Done.

**Comment by Andries-Smit** [comment]
First version will com from different branch

**Pull request closed without merging**